README
This is the public release of libelf-0.8.11, a free ELF object
file access library. If you have problems with applications
that use libelf and work with the commercial (SVR4, Solaris)
version but not with this one, please contact me.

IMPORTANT NOTE: If you have libelf-0.5.2 installed, you probably
have a file .../include/elf.h that contains the single line
``#include <libelf/elf.h>''. REMOVE THIS FILE BEFORE YOU RUN
configure.

Installation is straightforward - the package is autoconf'ed.  Just do
``cd libelf-0.8.11; ./configure; make; make install''.  Header files
will be installed in .../include/libelf/.  If your system does not
provide its own versions of libelf.h, nlist.h or gelf.h, ``make
install'' will add the missing headers.  If you prefer not to have
these files installed in /usr/include, use ``--disable-compat'' and
add ``-I /usr/include/libelf'' to your CFLAGS when compiling
libelf-based programs.

Note to distribution makers:  You can install libelf in a separate root
hierarchy by using the command ``make instroot=/my/root install''.
You should also use the ``--enable-compat'' configure option in that
case, or run ``make instroot=/my/root install-compat'' manually, to
install all the required header files.

If you are running Linux with libc 5.* as the default C library,
and you plan to use the 64-bit functions, you must either use
``-I.../include/libelf'', or remove /usr/include/libelf.h and use
``--enable-compat'' when running configure. Libc 6.* (aka glibc2)
doesn't have its own <libelf.h>, <nlist.h> or <gelf.h>.

You need an ANSI/ISO C compiler to build libelf. Gcc is optimal.

On some systems (in particular, Solaris and all variants of Linux),
``make'' will try to build a shared library. If you run into problems
on your system, please pass ``--disable-shared'' to configure.
If you build a shared library and want it to be installed as
``libelf-0.8.11.so'' rather than ``libelf.so.0.8.11'', please use
``./configure --enable-gnu-names''. Other files, e.g. ``libelf.so'' and
``libelf.so.0'' are NOT affected.

Another configure option, ``--enable-debug'', adds debugging code to
libelf; if you don't run into problems, you will probably not need it.

When creating an ELF shared library, it is possible to add references
to other shared libraries in the DYNAMIC section of the resulting
file. The make variable DEPSHLIBS contains a list of libraries to add.
It is set to ``-lc'' on Linux systems, and empty otherwise. To
override this setting, use something like ``make DEPSHLIBS="-la -lb"''.
For Linux, `-lc' is included automagically.

NLS is available and enabled by default.  To turn it off, pass the
``--disable-nls'' option to configure.

Libelf can use gettext or catgets for accessing message
catalogs.  If gettext is available AND is part of libc (i.e. not
in a separate library), it will be used. Otherwise, configure
will look for catgets.  If you have gettext in a separate
library and want to use it, you should pass the library's name
to configure, e.g. ``LIBS=-lintl ./configure''. Note that you
MUST link your libelf-based applications with -lintl then,
which is probably not what you want, or change the DEPSHLIBS variable
described above (in case you're building a shared library).

If you have GNU gettext 0.10 installed on your system, and if GNU gettext
runs on top of the catgets interface (rather old Linux systems, using
libc5), configure will refuse to use it and use catgets instead. If you
absolutely want to use GNU gettext, go ahead and rebuild it (which is
IMHO a good idea in general in this case):

	cd .../gettext-0.10
	ac_cv_func_catgets=no ac_cv_func_gettext=no ./configure
	make
	make install

After that, return to the libelf build directory, remove
config.cache, and start over.

*** Large File Support (LFS) applications ***

Some 32-bit systems support files that are larger than the address space
of the architecture.  On these, the `off_t' data type may have 32 or
64 bits, depending on the API you choose.  Since off_t is also part of
the libelf API, in particular the Elf_Data and Elf_Arhdr structures,
an application compiled with large file support will need a version of
libelf that has also been compiled with LFS; otherwise, it won't work
correctly.  Similarly, a program compiled without LFS needs a library
compiled without LFS.

Note that libelf is currently unable to process large files on 32-bit
architectures, whether you compile it for LFS or not, for the simple
reason that the files won't fit into the processes' address space.
Therefore, libelf is compiled without LFS by default.  It can of course
read and write ELF files for 64-bit architectures, but they will be
limited in length on a 32-bit system.

You may compile libelf with large file support by setting CPPFLAGS at
configuration time:

	CPPFLAGS=`getconf LFS_CFLAGS` ./configure

But I really, really recommend you don't, because it breaks binary
compatibility with existing libelf based applications.

*** 64-bit support ***

Starting with libelf-0.7.0, libelf also supports 64-bit ELF files.
This is enabled by default unless your system (or your compiler) does
not support 64-bit integers, or lacks 64-bit declarations in <elf.h>.
If you have problems building with 64-bit support, please do

    ./configure --disable-elf64

for the moment, and contact me. Please note that I haven't tested 64-bit
support much.  There are still some unresolved problems, e.g. IRIX
uses different Elf64_Rel and Elf64_Rela structures (they replaced the
r_info member), and the enumeration values for Elf_Type differ from
the commercial (SVR4) implementation of libelf - they broke binary
compatibility for no good reason, and I'm not willing to follow their
footsteps. The result is that libelf-0.7.* ist upward compatible with
libelf-0.6.4 (as it should be) but INCOMPATIBLE WITH SVR4 LIBELF. If you
have both versions installed, you'd better make sure that you link with
the library that matches the <libelf.h> you're #include'ing.

*** Symbol Versioning ***

Libelf >= 0.8.0 supports the data structures and definitions used for
symbol versioning on Solaris and Linux, in particular, the Elfxx_Verdef,
Elfxx_Verdaux, Elfxx_Verneed, Elfxx_Vernaux and Elfxx_Versym structures
and the SHT_XXX_verdef, SHT_XXX_verneed and SHT_XXX_versym section types
(where `xx' is either `32' or `64', and `XXX' is either `SUNW' or `GNU').
Libelf now translates versioning sections to/from their external
representation properly (earlier versions left them in `raw' format,
with the data type set to ELF_T_BYTE).  This may cause problems on
systems which use the same (OS-specific) section types for different
purposes.  The configure program tries to figure out if your OS uses
versioning; if that check fails, you can use

    ./configure --disable-versioning

to turn off versioning translation support.

*** W32 Support ***

There is now some support for building on W32 systems (requires Microsoft
VC++).  In order to build a W32 DLL, cd into the `lib' subdirectory, edit
build.bat if necessary (it needs the path to your compiler binaries) and
run it.  If you're lucky, libelf.dll and the import/export libraries will
be built.  If not, please drop me a line.

I tested it on XP Pro (SP2), using VC++ 2005 Express Edition.
Apparently, Visual Studio .NET 2003 works fine as well.

Various notes regarding the W32 port:

    - When you open() an ELF file, remember to use the O_BINARY flag.
    - You may have to add /MD to the linker command line.

*** Missing things ***

	* There is no documentation.  You can use the Solaris
	  manpages instead (available at http://docs.sun.com/).
	  The ELF file format is described in several places;
	  among them Suns "Linker and Libraries Guide" and the
	  "System V Application Binary Interface" documents;
	  http://www.caldera.com/developer/devspecs/gabi41.pdf and
	  http://www.caldera.com/developer/gabi/ are probably good
	  starting points. Processor-specific documentation is spread
	  across a number of `Processor Supplement' documents, one
	  for each architecture; you'll have to use a search engine to
	  find them.

	* The COFF file format is not understood. This is so obsolete
	  that it will probably never be implemented.

	* nlist(3) is incomplete; the n_type and n_sclass
	  members of struct nl are set to zero even if type
	  (that is, debug) information is available.

	* Libelf does not translate Solaris' `Move' and `Syminfo'
	  sections. You can read them using elf_getdata(), but you'll
	  only get raw (untranslated) bytes.

Changes since 0.8.10:

	* Fixed a bug in elf_rawfile().
	* If you use ELF_F_LAYOUT together with ELF_F_LAYOUT_OVERLAP,
	  elf_update() will now tolerate overlapping sections.

Changes since 0.8.9:

	* Ported to QNX Neutrino.
	* Fixed Windows build errors.
	* Parallel (make -j) installation should work now.

	* It's now possible to enable and disable select sanity checks
	  libelf performs. Currently, this affects the "NUL terminated
	  string table entry" check performed in elf_strptr(). By
	  default, the function will return an error if the string
	  requested is not properly terminated - because some
	  applications might dump core otherwise. If you configure
	  libelf with `--disable-sanity-checks', however, the check
	  (and, in the future, probably others as well) is disabled
	  by default. You can still turn it on and off at runtime by
	  setting the LIBELF_SANITY_CHECKS environment variable to
	  an integer value:

	      # disable all sanity checks
	      export LIBELF_SANITY_CHECKS=0

	      # enable all sanity checks
	      export LIBELF_SANITY_CHECKS=-1

	  Each bit of the value corresponds to a particular check,
	  so you could use LIBELF_SANITY_CHECKS=1 to enable only
	  the elf_strptr() check. You may also use a value in hex
	  (0x...) or octal (0...) format.
 
Changes since 0.8.8:

	* Improved translator for symbol versioning sections.
	* The W32 library is now built in the `lib' subdirectory.
	* Windows DLLs should work now.

Changes since 0.8.6:

	* added elf_getphnum().
	* added elf_getshnum().
	* added elf_getshstrndx().
	* added elfx_update_shstrndx().
	* handle interrupted reads/writes more gracefully.
	* added (partial) support for unusual e_[ps]hentsize values.
	* fixed the bugs introduced in 0.8.7.

Changes since 0.8.5:

	* added W32 support.
	* added workaround for alignment errors in archive members.
	* my email address has changed again ;)

Changes since 0.8.4:

	* elf_strptr() should now work more safely with fragmented
	  or badly formatted string tables.

Changes since 0.8.3:

	* Fixed a bug in elf_update() that was introduced in 0.8.3.

Changes since 0.8.2:

	* Should compile on MacOSX now.

	* Can read and write files with more than 65280 sections

	* Tries to handle 64-bit ELF files that use 8-byte hash table
	  entries. In particular, libelf tries to guess the data type in
	  elf_getdata(), and doesn't override sh_entsize in elf_update()
	  any longer. If you want the library to pick the entry size,
	  you must set its value to 0 before you call elf_update().

	* No longer dumps core in elf_update() when a versioning section
	  has no data. Instead, it returns an error message. Note that
	  you're supposed to provide a valid d_buf for any section, unless
	  it's empty or has SHT_NOBITS type.

	* Building a shared library is now the default (if supported).

Changes since 0.8.0:

	* Corrected typo in lib/{32,64}.xlatetof.c that sometimes
	  caused a compilation failure.

	* Use open(name, O_RDONLY|O_BINARY) in lib/nlist.c.

Changes since 0.7.0:

	* I implemented the gelf_* interface, as found on Solaris.
	  I don't know whether it's compatible -- the Solaris manpage
	  isn't very specific, so I had to guess return values etc. in
	  some cases.

	* Added elf{32,64}_checksum (supposed to be compatible with
	  Solaris).

	* Added symbol versioning support.

Changes since 0.6.4:

	* Fixed configure for IRIX systems
	* Added check for truncated archive members
	* Added check for misaligned SHDR/PHDR tables
	* Support for building libelf together with GNU libc
	* Added elf_memory(3)
	* Added 64-bit support

Changes since 0.5.2:

	* some bug fixes
	* mmap support
	* new directory layout
	* There is a new function, elf_delscn(), that deletes
	  a section from an ELF file. It also adjusts the
	  sh_link and sh_info members in the section header
	  table, if (and ONLY if) the ELF standard indicates
	  that these values are section indices. References
	  to the deleted section will be cleared, so be careful.
	* my email address has changed ;)

Where to get libelf:

	ftp://ftp.ibiblio.org/pub/Linux/libs/
	http://www.mr511.de/software/

Michael "Tired" Riepe
<libelf@mr511.de>


ChangeLog
Sat Jul  8 00:17:00 CEST 2006, Michael Riepe <libelf@mr511.de>

	* lib/32.newehdr.c:
		make return value compatible with Solaris.
	* lib/32.newphdr.c:
		handle 65535+ segments.
		make return value compatible with Solaris.
	* lib/cook.c:
		handle 65535+ segments.
	* lib/elf_repl.h:
		add new definitions.
	* lib/libelf.h:
		add/rename functions.
	* lib/newscn.c:
		fix section creation (was broken in 0.8.7).
	* lib/private.h:
		add SHN_XINDEX and PN_XNUM in case they're missing.
		centrally define LIBELF_SUCCESS and LIBELF_FAILURE.
	* lib/update.c:
		handle 65535+ segments.
		use elf->e_phnum internally.
	* lib/x.elfext.c:
		add elf_getphnum().
		rename elfx_get_shnum() and elfx_get_shstrndx().
		make return values compatible with Solaris.

Fri Jul  7 19:01:04 CEST 2006, Michael Riepe <libelf@mr511.de>

	* VERSION:
		bump up to 0.8.8.

Fri Jul  7 18:27:25 CEST 2006, Michael Riepe <libelf@mr511.de>

	* lib/Makefile.in:
		add lib/x.elfext.c.
	* lib/libelf.h:
		add functions from lib/x.elfext.c.
	* lib/newscn.c:
		simplify _elf_update_shnum().

Tue Apr 25 16:26:39 CEST 2006, Michael Riepe <libelf@mr511.de>

	* lib/gelf.h:
	* lib/libelf.h:
	* lib/nlist.h:
	* lib/private.h:
		add workaround for broken compilers.

Mon Apr 24 16:24:32 CEST 2006, Michael Riepe <libelf@mr511.de>

	* po/de.po:
		update.

Fri Apr 21 19:17:46 CEST 2006, Michael Riepe <libelf@mr511.de>

	* acconfig.h:
	* configure.in:
		add --enable-extended-format.
	* aclocal.m4:
		search for msgmerge.
	* lib/cook.c:
		change _elf_item buffering.
		handle extended format (with unusual entsize).
	* lib/errors.h:
	 	add ERROR_EHDR_SHENTSIZE and ERROR_EHDR_PHENTSIZE.
	* po/Makefile.in:
		use msgmerge instead of tupdate.
 
Thu Oct 20 21:08:02 CEST 2005, Michael Riepe <libelf@mr511.de>

	* lib/input.c:
	* lib/update.c:
		handle partial reads and writes.

Tue Aug 16 01:48:17 CEST 2005, Michael Riepe <libelf@mr511.de>

	* lib/begin.c:
		add workaround for archive member misalignment.
	* VERSION:
		bump up to 0.8.7

Tue Jul 19 11:56:26 CEST 2005, Michael Riepe <libelf@mr511.de>

	* README:
	* w32/build.bat:
		update.
	* w32/libelf.def:
		fix syntax.

Tue Jun 28 00:31:24 CEST 2005, Michael Riepe <libelf@mr511.de>

	* Makefile.in:
		remove superfluous slash.

Tue Jun 21 03:58:47 CEST 2005, Michael Riepe <libelf@mr511.de>

	* lib/Makefile.in:
		get rid of lib/pic subdir.

Sat May 21 17:39:28 CEST 2005, Michael Riepe <libelf@mr511.de>

	* (global):
		remove my e-mail address from all copyright clauses.

Sun May 15 23:08:30 CEST 2005, Michael Riepe <libelf@mr511.de>

	* configure.in:
		check if $CC can copile <elf.h>.
	* lib/private.h:
		#include <stdint.h> before <sys/types.h> (fixes glibc bug).

Sun May  8 23:40:35 CEST 2005, Michael Riepe <libelf@mr511.de>

	* Makefile.in:
		add instroot variable.
		install libelf.pc.
	* configure.in:
		create libelf.pc.

Sun Mar 20 15:41:22 CET 2005, Michael Riepe <libelf@mr511.de>

	* (global):
		change my e-mail address.

Fri Jan 28 23:09:57 CET 2005, Michael Riepe <michael@stud.uni-hannover.de>

	* po/Makefile.in:
		use modified gmo2msg.
	* po/gmo2msg.c:
		make gmo2msg output more portable.

Thu Oct  7 11:37:09 CEST 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* lib/cook.c:
		only use Elf64_Shdr if __LIBELF64 is true.

Fri Sep 17 02:55:47 CEST 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* lib/elf_repl.h:
		add some ABI and architecture definitions.
	* w32/config.h:
		manual update.

Sat Jul 10 17:33:15 CEST 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* acconfig.h:
	* aclocal.m4:
	* lib/errmsg.c:
		check for dgettext, not for gettext.
	* configure.in:
		check for -lintl.
	* po/Makefile.in:
		use -lintl when building gmo2msg.

Sun Jul  4 23:57:21 CEST 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* Makefile.in:
		add w32 subdir.
	* README:
		update for 0.8.6.
	* configure.in:
		create w32/Makefile.

Sat Jul  3 20:42:00 CEST 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* lib/32.xlatetof.c:
	* lib/64.xlatetof.c:
	* lib/Makefile.in:
		give up on <libelf_u.h>.
	* lib/getarsym.c:

Wed Jun 23 01:07:46 CEST 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* config.guess:
	* config.sub:
		update from FSF.

Tue May  4 22:02:01 CEST 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* config.guess:
	* config.sub:
		update from FSF.

Tue Mar 30 15:09:00 CEST 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* lib/32.xlatetof.c:
	* lib/64.xlatetof.c:
	* lib/Makefile.in:
		use <libelf_u.h> to work around W32 compiler problems.

Mon Feb 16 06:19:11 CET 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* Makefile.in:
		generate old-format tar file.

Sat Jan 24 03:42:39 CET 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* lib/32.xlatetof.c:
	* lib/64.xlatetof.c:
		replace NULL with 0 -- some compilers don't like (void*).
	* lib/getarsym.c:
	* lib/nlist.c:
		add cast to suppress compiler warning.

Fri Jan 23 05:11:46 CET 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* lib/update.c:
		#undef max before #define.

Wed Jan 21 18:15:50 CET 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* lib/begin.c:
		better support for Cygwin .lib archive files.

Mon Jan 19 15:36:21 CET 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* lib/libelf.h:
	* lib/memset.c:
		include <stddef.h> unconditionally.

Fri Jan 16 23:13:25 CET 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* aclocal.m4:
		support Intel C Compiler.
	* lib/32.newehdr.c:
	* lib/32.newphdr.c:
		remove elf->e_free_ehdr and elf->e_free_phdr.
	* lib/cook.c:
		always allocate ehdr and phdr.
	* lib/end.c:
		always deallocate ehdr and phdr.
	* lib/private.h:
		remove elf->e_free_ehdr and elf->e_free_phdr.
		change valid_type to suppress compiler warning.
	* lib/update.c:
		not necessary to update elf->e_ehdr and elf->e_phdr.

Thu Jan 15 22:43:00 CET 2004, Michael Riepe <michael@stud.uni-hannover.de>

	* VERSION:
		bump up to 0.8.6.
	* configure.in:
		check for __int64.
	* lib/32.xlatetof.c:

lib/libelf.h
/*
 * libelf.h - public header file for libelf.
 * Copyright (C) 1995 - 2008 Michael Riepe
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
 */

/* @(#) $Id: libelf.h,v 1.28 2009/05/22 17:08:09 michael Exp $ */

#ifndef _LIBELF_H
#define _LIBELF_H

#include <stddef.h>	/* for size_t */
#include <sys/types.h>

#if __LIBELF_INTERNAL__
#include <sys_elf.h>
#else /* __LIBELF_INTERNAL__ */
#include <libelf/sys_elf.h>
#endif /* __LIBELF_INTERNAL__ */

#ifdef __cplusplus
extern "C" {
#endif /* __cplusplus */

#ifndef __P
# if (__STDC__ + 0) || defined(__cplusplus) || defined(_WIN32)
#  define __P(args) args
# else /* __STDC__ || defined(__cplusplus) */
#  define __P(args) ()
# endif /* __STDC__ || defined(__cplusplus) */
#endif /* __P */

/*
 * Commands
 */
typedef enum {
    ELF_C_NULL = 0,	/* must be first, 0 */
    ELF_C_READ,
    ELF_C_WRITE,
    ELF_C_CLR,
    ELF_C_SET,
    ELF_C_FDDONE,
    ELF_C_FDREAD,
    ELF_C_RDWR,
    ELF_C_NUM		/* must be last */
} Elf_Cmd;

/*
 * Flags
 */
#define ELF_F_DIRTY	0x1
#define ELF_F_LAYOUT	0x4
/*
 * Allow sections to overlap when ELF_F_LAYOUT is in effect.
 * Note that this flag ist NOT portable, and that it may render
 * the output file unusable.  Use with extreme caution!
 */
#define ELF_F_LAYOUT_OVERLAP	0x10000000

/*
 * File types
 */
typedef enum {
    ELF_K_NONE = 0,	/* must be first, 0 */
    ELF_K_AR,
    ELF_K_COFF,
    ELF_K_ELF,
    ELF_K_NUM		/* must be last */
} Elf_Kind;

/*
 * Data types
 */
typedef enum {
    ELF_T_BYTE = 0,	/* must be first, 0 */
    ELF_T_ADDR,
    ELF_T_DYN,
    ELF_T_EHDR,
    ELF_T_HALF,
    ELF_T_OFF,
    ELF_T_PHDR,
    ELF_T_RELA,
    ELF_T_REL,
    ELF_T_SHDR,
    ELF_T_SWORD,
    ELF_T_SYM,
    ELF_T_WORD,
    /*
     * New stuff for 64-bit.
     *
     * Most implementations add ELF_T_SXWORD after ELF_T_SWORD
     * which breaks binary compatibility with earlier versions.
     * If this causes problems for you, contact me.
     */
    ELF_T_SXWORD,
    ELF_T_XWORD,
    /*
     * Symbol versioning.  Sun broke binary compatibility (again!),
     * but I won't.
     */
    ELF_T_VDEF,
    ELF_T_VNEED,
    ELF_T_NUM		/* must be last */
} Elf_Type;

/*
 * Elf descriptor
 */
typedef struct Elf	Elf;

/*
 * Section descriptor
 */
typedef struct Elf_Scn	Elf_Scn;

/*
 * Archive member header
 */
typedef struct {
    char*		ar_name;
    time_t		ar_date;
    long		ar_uid;
    long 		ar_gid;
    unsigned long	ar_mode;
    off_t		ar_size;
    char*		ar_rawname;
} Elf_Arhdr;

/*
 * Archive symbol table
 */
typedef struct {
    char*		as_name;
    size_t		as_off;
    unsigned long	as_hash;
} Elf_Arsym;

/*
 * Data descriptor
 */
typedef struct {
    void*		d_buf;
    Elf_Type		d_type;
    size_t		d_size;
    off_t		d_off;
    size_t		d_align;
    unsigned		d_version;
} Elf_Data;

/*
 * Function declarations
 */
extern Elf *elf_begin __P((int __fd, Elf_Cmd __cmd, Elf *__ref));
extern Elf *elf_memory __P((char *__image, size_t __size));
extern int elf_cntl __P((Elf *__elf, Elf_Cmd __cmd));
extern int elf_end __P((Elf *__elf));
extern const char *elf_errmsg __P((int __err));
extern int elf_errno __P((void));
extern void elf_fill __P((int __fill));
extern unsigned elf_flagdata __P((Elf_Data *__data, Elf_Cmd __cmd,
	unsigned __flags));
extern unsigned elf_flagehdr __P((Elf *__elf, Elf_Cmd __cmd,
	unsigned __flags));
extern unsigned elf_flagelf __P((Elf *__elf, Elf_Cmd __cmd,
	unsigned __flags));
extern unsigned elf_flagphdr __P((Elf *__elf, Elf_Cmd __cmd,
	unsigned __flags));
extern unsigned elf_flagscn __P((Elf_Scn *__scn, Elf_Cmd __cmd,
	unsigned __flags));
extern unsigned elf_flagshdr __P((Elf_Scn *__scn, Elf_Cmd __cmd,
	unsigned __flags));
extern size_t elf32_fsize __P((Elf_Type __type, size_t __count,
	unsigned __ver));
extern Elf_Arhdr *elf_getarhdr __P((Elf *__elf));
extern Elf_Arsym *elf_getarsym __P((Elf *__elf, size_t *__ptr));
extern off_t elf_getbase __P((Elf *__elf));
extern Elf_Data *elf_getdata __P((Elf_Scn *__scn, Elf_Data *__data));
extern Elf32_Ehdr *elf32_getehdr __P((Elf *__elf));
extern char *elf_getident __P((Elf *__elf, size_t *__ptr));
extern Elf32_Phdr *elf32_getphdr __P((Elf *__elf));
extern Elf_Scn *elf_getscn __P((Elf *__elf, size_t __index));
extern Elf32_Shdr *elf32_getshdr __P((Elf_Scn *__scn));
extern unsigned long elf_hash __P((const unsigned char *__name));
extern Elf_Kind elf_kind __P((Elf *__elf));
extern size_t elf_ndxscn __P((Elf_Scn *__scn));
extern Elf_Data *elf_newdata __P((Elf_Scn *__scn));
extern Elf32_Ehdr *elf32_newehdr __P((Elf *__elf));
extern Elf32_Phdr *elf32_newphdr __P((Elf *__elf, size_t __count));
extern Elf_Scn *elf_newscn __P((Elf *__elf));
extern Elf_Cmd elf_next __P((Elf *__elf));
extern Elf_Scn *elf_nextscn __P((Elf *__elf, Elf_Scn *__scn));
extern size_t elf_rand __P((Elf *__elf, size_t __offset));
extern Elf_Data *elf_rawdata __P((Elf_Scn *__scn, Elf_Data *__data));
extern char *elf_rawfile __P((Elf *__elf, size_t *__ptr));
extern char *elf_strptr __P((Elf *__elf, size_t __section, size_t __offset));
extern off_t elf_update __P((Elf *__elf, Elf_Cmd __cmd));
extern unsigned elf_version __P((unsigned __ver));
extern Elf_Data *elf32_xlatetof __P((Elf_Data *__dst, const Elf_Data *__src,
	unsigned __encode));
extern Elf_Data *elf32_xlatetom __P((Elf_Data *__dst, const Elf_Data *__src,
	unsigned __encode));

/*
 * Additional functions found on Solaris
 */
extern long elf32_checksum __P((Elf *__elf));

#if __LIBELF64
/*
 * 64-bit ELF functions
 * Not available on all platforms
 */
extern Elf64_Ehdr *elf64_getehdr __P((Elf *__elf));
extern Elf64_Ehdr *elf64_newehdr __P((Elf *__elf));
extern Elf64_Phdr *elf64_getphdr __P((Elf *__elf));
extern Elf64_Phdr *elf64_newphdr __P((Elf *__elf, size_t __count));
extern Elf64_Shdr *elf64_getshdr __P((Elf_Scn *__scn));
extern size_t elf64_fsize __P((Elf_Type __type, size_t __count,
	unsigned __ver));
extern Elf_Data *elf64_xlatetof __P((Elf_Data *__dst, const Elf_Data *__src,
	unsigned __encode));
extern Elf_Data *elf64_xlatetom __P((Elf_Data *__dst, const Elf_Data *__src,
	unsigned __encode));

/*
 * Additional functions found on Solaris
 */
extern long elf64_checksum __P((Elf *__elf));

#endif /* __LIBELF64 */

/*
 * ELF format extensions
 *
 * These functions return 0 on failure, 1 on success.
 */
extern int elf_getphnum __P((Elf *__elf, size_t *__resultp));
extern int elf_getshnum __P((Elf *__elf, size_t *__resultp));
extern int elf_getshstrndx __P((Elf *__elf, size_t *__resultp));

/*
 * Convenience functions
 *
 * elfx_update_shstrndx is elf_getshstrndx's counterpart.
 * It should be used to set the e_shstrndx member.
 * There is no update function for e_shnum or e_phnum
 * because libelf handles them internally.
 */
extern int elfx_update_shstrndx __P((Elf *__elf, size_t __index));

/*
 * Experimental extensions:
 *
 * elfx_movscn() moves section `__scn' directly after section `__after'.
 * elfx_remscn() removes section `__scn'.  Both functions update
 * the section indices; elfx_remscn() also adjusts the ELF header's
 * e_shnum member.  The application is responsible for updating other
 * data (in particular, e_shstrndx and the section headers' sh_link and
 * sh_info members).
 *
 * elfx_movscn() returns the new index of the moved section.
 * elfx_remscn() returns the original index of the removed section.
 * A return value of zero indicates an error.
 */
extern size_t elfx_movscn __P((Elf *__elf, Elf_Scn *__scn, Elf_Scn *__after));
extern size_t elfx_remscn __P((Elf *__elf, Elf_Scn *__scn));

/*
 * elf_delscn() is obsolete.  Please use elfx_remscn() instead.
 */
extern size_t elf_delscn __P((Elf *__elf, Elf_Scn *__scn));

#ifdef __cplusplus
}
#endif /* __cplusplus */

#endif /* _LIBELF_H */
